File: 23. Check valid max-heap.cpp
/*
 Task: Check if a given array represents a valid max-heap
*/

#include <bits/stdc++.h>
using namespace std;

bool isMaxHeap(const vector<int>& a)
{
    int n = a.size();
    for (int i = 0; i <= (n - 2) / 2; ++i)
    {
        if (2 * i + 1 < n && a[i] < a[2 * i + 1])
            return false;
        if (2 * i + 2 < n && a[i] < a[2 * i + 2])
            return false;
    }
    return true;
}

int main()
{
    vector<int> a = {90, 15, 10, 7, 12, 2, 7, 3};
    cout << (isMaxHeap(a) ? "Yes" : "No") << "\n";
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)

File: 23. Check valid max-heap.py
"""
Task: Check if a given array represents a valid max-heap
Time Complexity: O(n)
Space Complexity: O(1)
"""

def is_max_heap(a):
    n=len(a)
    for i in range((n-2)//2+1):
        l=2*i+1; r=2*i+2
        if l<n and a[i] < a[l]: return False
        if r<n and a[i] < a[r]: return False
    return True

print(is_max_heap([90,15,10,7,12,2,7,3]))
